### ISSUE TITLE
Do not apply `UseIfInsteadOfWhen` when variable captured

### ISSUE THREAD

**Issue by EdwarDDay** [detekt/detekt#5568]
## Expected Behavior of the rule

Currently the `UseIfInsteadOfWhen` rule complains, that the following code is not compliant:
```kotlin
when (val x = calculateSomething()) {
    is Success -> doSomething(x)
    else -> println(x)
}
```
This would then need to be changed to something like
```kotlin
val x = calculateSomething()
if (x is Success) {
    doSomething(x)
} else {
    println(x)
}
```
I would suggest to not apply the rule to `when` with a captured variable, because the changed version would expose the variable to a greater scope (and not only the conditional block).
 
This could also be added as a configuration for the rule, so that it can be generally enabled, but disabled for the named cases.

## Context

Jetbrains removed the `replace 'when' with 'if'` [quick fix from Intellij](https://youtrack.jetbrains.com/issue/KTIJ-15573/Intention-Replace-when-with-if-produces-wrong-code-if-expression-subject-is-a-variable-declaration), when the `when` expression uses a captured variable. So this makes this rule also harder to apply in these cases.

**Comment by cortinico** [comment]
> This would then need to be changed to something like

That's an interesting case. I would argue that the second snippet is more readable though.
Something to raise to JetBrains as a language proposals for capturing variable inside if conditions maybe?

**Comment by EdwarDDay** [comment]
I think for an `if` block it would be not so clear, for which block the variable would be captured (only the if block or additional `else` or `else if` blocks?).

**Comment by cortinico** [comment]
> would be captured (only the if block or additional `else` or `else if` blocks?).

All the blocks in the same if hierarchy (`if`, all the `if else` and the `else`).
Back to your suggestion:

> I would argue that the second snippet is more readable though.

What do you think about this?

**Comment by EdwarDDay** [comment]
I wouldn't see a big difference in the readability. I think this mainly depends on the developer to decide, what is more readable in this case.

**Comment by BraisGabin** [comment]
I could argue that the first one reduces the scope of the variable and that's nice. So maybe adding that configuration is not a bad idea.

**Comment by atulgpt** [comment]
Hi, @BraisGabin / @cortinico  I have raised the [PR](https://github.com/detekt/detekt/pull/5681) for this. This PR is inspired by JetBrains Ide work at [this PR](https://github.com/JetBrains/kotlin/pull/2921)
